From d8de62ff2dc076edfbf60d994358a44da119b1eb Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 05 Nov 2021 13:07:24 +0200
Subject: [PATCH] [SW-0] gaudi2: scrub arc0 iccm memory on cold reset

In gaudi2, ARC0 ICCM is used to store boot info. Since boot info read
is performed first as part of preboot, fake ECC exception can be
observed due to uninitialized reads performed on ARC0 memory.
In order to avoid ECC exceptions due to uninitialized reads,
this patch add changes to scrub ARC0 ICCM memory in case of cold reset.

Change-Id: Id87d59d44b88386739dadf1c9ce2ad1ea45ea1f5
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index ce88ef7..351e35a 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -84,6 +84,8 @@
 
 void init_secondary(void)
 {
+	u32 rst_src;
+
 	halt_n_reset_arc(0);
 
 	/*
@@ -97,6 +99,13 @@
 
 	/* Clear DCCM memory */
 	init_arc_mem(0, ARC0_DCCM_FULL_BASE, ARC0_DCCM_SIZE);
+
+	/* Clear ARC0 ICCM memory on cold reset only */
+	rst_src = hl_rst_src_get();
+	if (rst_src != HL_COLD_RST)
+		return;
+
+	init_arc_mem(0, ARC0_ICCM_FULL_BASE, ARC0_ICCM_SIZE);
 }
 
 void hl_plat_early_init(void)
diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index 992dd92..6be8f53 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -156,6 +156,8 @@
 
 static void init_arc(void)
 {
+	u32 rst_src;
+
 	halt_n_reset_arc(0);
 	halt_n_reset_arc(1);
 
@@ -173,6 +175,13 @@
 	/* Clear DCCM memory */
 	init_arc_mem(0, ARC0_DCCM_FULL_BASE, ARC0_DCCM_SIZE);
 	init_arc_mem(1, ARC1_DCCM_FULL_BASE, ARC1_DCCM_SIZE);
+
+	/* Clear ARC0 ICCM memory on cold reset only */
+	rst_src = hl_rst_src_get();
+	if (rst_src != HL_COLD_RST)
+		return;
+
+	init_arc_mem(0, ARC0_ICCM_FULL_BASE, ARC0_ICCM_SIZE);
 }
 
 void hl_plat_init(void)
